src/dmd/escape.d

```diff
@@ -585,9 +585,142 @@ bool checkThrowEscape(Scope* sc, Expression e, bool gag)
 bool checkNewEscape(Scope* sc, Expression e, bool gag)
 {
     //printf("[%s] checkNewEscape, e = %s\n", e.loc.toChars(), e.toChars());
-    return checkReturnEscapeImpl(sc, e, false, gag, false);
+    enum log = false;
+    if (log) printf("[%s] checkNewEscape, e: `%s`\n", e.loc.toChars(), e.toChars());
+    EscapeByResults er;
+
+    escapeByValue(e, &er);
+
+    if (!er.byref.dim && !er.byvalue.dim && !er.byexp.dim)
+        return false;
+
+    bool result = false;
+    foreach (VarDeclaration v; er.byvalue)
+    {
+        if (log) printf("byvalue `%s`\n", v.toChars());
+        if (v.isDataseg())
+            continue;
+
+        Dsymbol p = v.toParent2();
+
+        if (v.isScope())
+        {
+            if (sc._module && sc._module.isRoot() &&
+                /* This case comes up when the ReturnStatement of a __foreachbody is
+                 * checked for escapes by the caller of __foreachbody. Skip it.
+                 *
+                 * struct S { static int opApply(int delegate(S*) dg); }
+                 * S* foo() {
+                 *    foreach (S* s; S) // create __foreachbody for body of foreach
+                 *        return s;     // s is inferred as 'scope' but incorrectly tested in foo()
+                 *    return null; }
+                 */
+                !(p.parent == sc.func))
+            {
+                // Only look for errors if in module listed on command line
+                if (global.params.vsafe) // https://issues.dlang.org/show_bug.cgi?id=17029
+                {
+                    if (!gag)
+                        error(e.loc, "scope variable `%s` may not be copied into allocated memory", v.toChars());
```

Comment:
I suggest using the term "GC-allocated memory" to be more precise.  The problem isn't about memory (per se) or its allocation but about the lifetimes of objects under the jurisdiction of the GC.  I suggest something like:

> scope variable `%s` may not be copied into GC-allocated memory because doing so would cause it to escape

Reply:
I didn't use that term because the message could apply to any allocated memory.